firedrake/petsc.py

```diff
@@ -334,3 +370,70 @@ def garbage_view(obj: Any):
         PETSc.garbage_view(comm)
     else:
         raise FiredrakePETScError("No comm found, cannot view garbage")
+
+
+external_packages = get_external_packages()
+
+# Setup default partitioner
+# Manually define the priority until
+# https://petsc.org/main/src/dm/partitioner/interface/partitioner.c.html#PetscPartitionerGetDefaultType
+# is added to petsc4py
+partitioner_priority = ["parmetis", "ptscotch", "chaco"]
+for partitioner in partitioner_priority:
+    if partitioner in external_packages:
+        DEFAULT_PARTITIONER = partitioner
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(partitioner_priority)
+        + " found, defaulting to PETSc simple partitioner. This may not be optimal."
+    )
+    DEFAULT_PARTITIONER = "simple"
+
+# Setup default direct solver
+direct_solver_priority = ["mumps", "superlu_dist", "pastix"]
+for solver in direct_solver_priority:
+    if solver in external_packages:
+        DEFAULT_DIRECT_SOLVER = solver
+        DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": solver}
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(direct_solver_priority)
+        + " found, defaulting to PETSc LU. This will only work in serial."
+    )
+    DEFAULT_DIRECT_SOLVER = "petsc"
+    DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": "petsc"}
+
+# MUMPS needs an additional parameter set
+# From the MUMPS documentation:
+# > ICNTL(14) controls the percentage increase in the estimated working space...
+# > ... Remarks: When significant extra fill-in is caused by numerical pivoting, increasing ICNTL(14) may help.
+if DEFAULT_DIRECT_SOLVER == "mumps":
+    DEFAULT_DIRECT_SOLVER_PARAMETERS["mat_mumps_icntl_14"] = 200
+
+# Setup default AMG preconditioner
+amg_priority = ["hypre", "ml"]
+for amg in amg_priority:
+    if amg in external_packages:
+        DEFAULT_AMG_PC = amg
+else:
+    DEFAULT_AMG_PC = "gamg"
+
+
```

Comment:
```suggestion
# The default parameter dictionaries must always be flattened to avoid accidental mutation when copying.
# Because of how python naming works, if DEFAULT_KSP_PARAMETERS is not flattened and
# DEFAULT_KSP_PARAMETERS.copy() is put into another dictionary, then the nested parts of
# DEFAULT_KSP_PARAMETERS could be mutated from the copy.
```

Reply:
This suggested comment isn't actually the reason we need to flatten the parameters. The issue is actually more subtle than that:

The function `set_defaults` in `solving_utils.py` mutates the dictionaries used in solvers based only on flattened keys, so if the `DEFAULT` dictionaries here are nested: ie:
```python
DEFAULT_KSP_PARAMETERS = {
    "mat_type": "aij",
    "ksp_type": "preonly",
    "ksp_rtol": 1e-7,
    "pc_type": "lu",
    "pc_factor": DEFAULT_DIRECT_SOLVER_PARAMETERS
}
```
Then trying to changing `"pc_factor_mat_solver_type"` in `set_defaults` fails and you see duplicated keys in PETSc options.

Really what is needed is a refactor of `set_defaults` as there is a _lot_ going on in there! However, I'm not currently willing to unpick that particular knot! Not with all the other changes going on here. Feel free to open an issue to document this though.

I will add a comment explaining this.